[API] Implementation of Achievement Creation Endpoint

abrodriguezalman opened the pull request:
In partnership with the Achievements team, this issue aims to develop an API endpoint to allow approved users* to create ACHIEVEMENTS for a given GAME. This PR:
1. Creates a serializer for the Achievement model created by the Achievements team
2. Implements a new function AchievementCreateView() under api/views.py with a corresponding URL path "api/games/game_id/achievements/create/".

These two modifications allow users to create achievements in the database for a given game.

Commit:
[API]Finished implementation of Achievement creation endpoint

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-create-endpoint

Commit:
[API] Added success message to be returned after sucessful POST request in AchievementCreateView()

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-create-endpoint

chrislowzhengxi reviewed (approved):
LGTM. 
I checked the endpoint manually with a POST request and got a 201 CREATED with the expected response structure. 

I went to http://127.0.0.1:8000/api/games/1/achievements/create/ with 

{
  "name": "Sharp Shooter",
  "description": "Hit 100 targets",
  "rarity": 1,
  "threshold": 100
}

MuyanXie commented:
I think everything looks good. The only concern I have is that the intergrity error (when achievements have same names) are casted as a 500 error. Please make it a 400 error

MuyanXie reviewed (changes requested):
Encountered this error:
ntegrityError at /api/games/1/achievements/create/
UNIQUE constraint failed: achievements_achievement.name, achievements_achievement.game_id

when posting the achievement with the same name; Try to make it cast a 400 bad request http response, instead of a 500 error

Commit:
[API] Modified AchievementCreateView() to send 400 error message to meet the unique constraint between achievement name and achievement game

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-create-endpoint

abrodriguezalman commented:
> Encountered this error: ntegrityError at /api/games/1/achievements/create/ UNIQUE constraint failed: achievements_achievement.name, achievements_achievement.game_id
> 
> when posting the achievement with the same name; Try to make it cast a 400 bad request http response, instead of a 500 error

I have updated the AchievementCreateView() function so that a 400 error response is sent whenever a user tries to make an achievement with a name that already exists for a particular game. I also added a small error message to indicate the source of the problem to users.

MuyanXie reviewed (approved):
LGTM. I'm able to verify the changes

abrodriguezalman commented:
Concluding statement
In order to address this issue, the following changes were implemented to the codebase:

Creates a serializer for the Achievement model created by the Achievements team
Implements a new function AchievementCreateView() under api/views.py with a corresponding URL path "api/games/game_id/achievements/create/".

These two modifications allow users to create achievements in the database for a given game.

MuyanXie commented:
Issue Score: excellent

Comments:
Nice job! The implementation is very structured and the endpoints work well!

<!-- {"score": "excellent", "minor": false} -->
